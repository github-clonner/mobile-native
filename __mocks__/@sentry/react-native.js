export const captureMessage = jest.fn();
export default {};
