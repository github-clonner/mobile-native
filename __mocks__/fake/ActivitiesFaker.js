
function load(count, container) {

  let activities = [...Array(count)].map((_, i) => {
    const code = 'activityguid' + i;
    return {
      attachment_guid: false,
      blurb: false,
      container_guid:code,
      custom_data:false,
      custom_type:false,
      rowKey: 'something' + i,
      description:"Congratulations! ",
      edited:"",
      guid:code,
      mature:false,
      time_created: "1522036284",
      ownerObj:{
        guid: "824853017709780997",
        type: "user",
        subtype: false,
        time_created: "1522036284",
        // getAvatarSource: () => {
        //   return {
        //     source:'http://thisisaurl'
        //   }
        // }
      },
      shouldBeBlured: jest.fn(),
      message:"Message",
      title:'TITLE',
      owner_guid:"824853017709780997",
      parent_guid:"838106762591510528",
      perma_url:false,
      thumbnail_src:false,
      type:"activity",
      wire_totals: {
        tokens: 1000000000000000000
      },
      containerObj: container
      // _list: {
      //   viewed: {
      //     viewed: new Map([["1019155171608096768",true]]),
      //     addViewed: () => {
      //       return;
      //     }
      //   }
      // },
      // getThumbSource: () => {
      //   return {
      //     source:'http://thisisaurl'
      //   }
      // }
    }
  });

  return {
    'activities': activities,
    'load-next': 'aaaaaa',
    'load-previous': 'aaaaaa'
  }

}

/**
 * comment service data faker
 * @param {integer} count
 */
export function activitiesServiceFaker() {
  return { load: load}
}

