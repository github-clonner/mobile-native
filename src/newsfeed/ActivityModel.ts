import { runInAction, action, observable, decorate } from 'mobx';
import FastImage from 'react-native-fast-image';
import { FlatList, Alert } from 'react-native';

import BaseModel from '../common/BaseModel';
import UserModel from '../channel/UserModel';
import wireService from '../wire/WireService';
import sessionService from '../common/services/session.service';
import {
  setPinPost,
  deleteItem,
  unfollow,
  follow,
  update,
} from '../newsfeed/NewsfeedService';
import api from '../common/services/api.service';

import { GOOGLE_PLAY_STORE, MINDS_CDN_URI, MINDS_URI } from '../config/Config';
import i18n from '../common/services/i18n.service';
import logService from '../common/services/log.service';
import entitiesService from '../common/services/entities.service';
import type { ThumbSize } from '../types/Common';
import type GroupModel from 'src/groups/GroupModel';

type Thumbs = Record<ThumbSize, string>;

/**
 * Activity model
 */
export default class ActivityModel extends BaseModel {
  // Observable properties
  @observable pinned: boolean = false;
  @observable time_created: string = '';
  @observable message: string = '';
  @observable title: string = '';
  @observable mature: boolean = false;
  @observable edited: '0' | '1' = '0';

  // decorated observables
  'is:following': boolean;
  'thumbs:down:count': number;
  'thumbs:up:count': number;
  'comments:count': number;
  'thumbs:down:user_guids': Array<number>;
  'thumbs:up:user_guids': Array<number>;

  containerObj?: GroupModel;
  remind_object?: ActivityModel;
  ownerObj!: UserModel;
  listRef?: FlatList<any>;
  thumbnails?: Thumbs;
  paywall: '1' | '' = '';
  paywall_unlocked: boolean = false;
  guid: string = '';
  subtype: string = '';
  entity_guid: string = '';
  owner_guid: string = '';
  custom_type?: string;
  custom_data?: Array<any> | any;
  nsfw?: Array<number>;
  flags?: any;
  reminds: number = 0;
  impressions: number = 0;
  perma_url?: string;
  cinemr_guid?: string;
  thumbnail_src?: string;
  dontPin?: boolean;
  boosted?: boolean;
  wire_threshold?: {
    type: 'token' | 'money';
    min: number;
  } | null;
  _preview?: boolean;
  /**
   * Set programatically to indicate that the parent is marked as mature
   */
  is_parent_mature?: boolean;

  /**
   * Mature visibility flag
   */
  @observable mature_visibility: boolean = false;

  /**
   * Is visible in flat list
   */
  @observable is_visible: boolean = true;

  /**
   *  List reference setter
   */
  set _list(value) {
    this.__list = value;

    // the reminded object need access to the metadata service too
    if (this.remind_object) {
      this.remind_object._list = value;
    }
  }

  /**
   *  List reference getter
   */
  get _list() {
    return this.__list;
  }

  toPlainObject() {
    const plainEntity = super.toPlainObject();
    if (plainEntity.remind_object && plainEntity.remind_object.__list) {
      delete plainEntity.remind_object.__list;
    }

    return plainEntity;
  }

  /**
   * Child models
   */
  childModels() {
    return {
      ownerObj: UserModel,
      remind_object: ActivityModel,
    };
  }

  /**
   * Get the activity thumb source
   * {uri: 'http...'}
   * @param {string} size
   */
  getThumbSource(size: ThumbSize = 'medium') {
    // for gif use always the same size to take adventage of the cache (they are not resized)
    if (this.isGif()) {
      size = 'medium';
    }

    if (this.thumbnails && this.thumbnails[size]) {
      return { uri: this.thumbnails[size], headers: api.buildHeaders() };
    }

    // fallback to old behavior
    if (this.paywall || this.paywall_unlocked) {
      return {
        uri: MINDS_URI + 'fs/v1/thumbnail/' + this.entity_guid,
        headers: api.buildHeaders(),
      };
    }
    if (this.custom_type === 'batch') {
      return {
        uri: MINDS_CDN_URI + 'fs/v1/thumbnail/' + this.entity_guid + '/' + size,
        headers: api.buildHeaders(),
      };
    }
    return {
      uri: MINDS_CDN_URI + 'fs/v1/thumbnail/' + this.guid + '/' + size,
      headers: api.buildHeaders(),
    };
  }

  isGif(): boolean {
    return this.custom_data && this.custom_data[0] && this.custom_data[0].gif;
  }

  /**
   * Preload thumb on image cache
   */
  preloadThumb(size: ThumbSize = 'medium') {
    FastImage.preload([this.getThumbSource(size)]);
  }

  shouldBeBlured(): boolean {
    const user = sessionService.getUser();

    if (user && user.mature) {
      return false;
    }

    if (this.nsfw !== undefined) {
      let res = [1, 2, 4].filter((nsfw) => this.nsfw!.indexOf(nsfw) > -1)
        .length;
      if (res) {
        return true;
      }
    }

    if (typeof this.flags !== 'undefined') {
      return !!this.flags.mature;
    }

    if (typeof this.mature !== 'undefined') {
      return !!this.mature;
    }

    if (
      typeof this.custom_data !== 'undefined' &&
      typeof this.custom_data[0] !== 'undefined'
    ) {
      return !!this.custom_data[0].mature;
    }

    if (this.custom_data !== undefined) {
      return !!this.custom_data.mature;
    }

    return false;
  }

  /**
   * Get activity text
   */
  get text() {
    return this.message || '';
  }

  @action
  toggleMatureVisibility() {
    if (GOOGLE_PLAY_STORE) {
      return;
    }
    this.mature_visibility = !this.mature_visibility;

    if (this.remind_object && this.remind_object.mature) {
      this.remind_object.mature_visibility = this.mature_visibility;
    }
  }

  @action
  setVisible(value: boolean) {
    this.is_visible = value;
  }

  /**
   * Increment the comments counter
   */
  @action
  incrementCommentsCounter() {
    this['comments:count']++;
  }

  /**
   * Decrement the comments counter
   */
  @action
  decrementCommentsCounter() {
    this['comments:count']--;
  }

  /**
   * Unlock the activity and update data on success
   */
  @action
  async unlock(ignoreError = false) {
    try {
      const result: object | false = await wireService.unlock(this.guid);
      if (result) {
        // all changes should be atomic (trigger render only once)
        runInAction(() => {
          // create a new model because we need the child models
          const model: ActivityModel = ActivityModel.create(result);
          Object.assign(this, model);
        });
      }
      return result;
    } catch (err) {
      if (!ignoreError) {
        Alert.alert(err.message);
      }
      return false;
    }
  }

  @action
  async togglePin() {
    // allow owners only
    if (!this.isOwner()) {
      return;
    }

    try {
      this.pinned = !this.pinned;
      await setPinPost(this.guid, this.pinned);
    } catch (e) {
      this.pinned = !this.pinned;
      Alert.alert(i18n.t('errorPinnedPost'));
    }
  }

  @action
  async deleteEntity() {
    try {
      await deleteItem(this.guid);
      this.removeFromList();
      entitiesService.deleteFromCache(this.urn);
    } catch (err) {
      logService.exception('[ActivityModel]', err);
      throw err;
    }
  }

  @action
  async toggleFollow() {
    const method = this['is:following'] ? unfollow : follow;
    try {
      await method(this.guid);
      runInAction(() => {
        this['is:following'] = !this['is:following'];
      });
    } catch (err) {
      logService.exception('[OffsetFeedListStore]', err);
      throw err;
    }
  }

  @action
  async updateActivity(data: any = {}) {
    const entity: any = this.toPlainObject();

    if (data) {
      for (const field in data) {
        entity[field] = data[field];
      }
    }

    // call update endpoint
    await update(entity);

    // update instance properties
    this.update(data);

    this.setEdited(entity.message);
  }

  @action
  setEdited(message: string) {
    this.message = message;
    this.edited = '1';
  }
}

/**
 * Define model observables
 */
decorate(ActivityModel, {
  //@ts-ignore
  'thumbs:down:count': observable,
  'thumbs:up:count': observable,
  'comments:count': observable,
  'thumbs:down:user_guids': observable,
  'thumbs:up:user_guids': observable,
});
