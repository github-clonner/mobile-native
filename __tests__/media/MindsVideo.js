import 'react-native';
import React from 'react';
import MindsVideo from '../../src/media/MindsVideo';
import { activitiesServiceFaker } from '../../__mocks__/fake/ActivitiesFaker';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import ActivityModel from '../../src/newsfeed/ActivityModel';

jest.mock('react-native-video');
jest.mock('../../src/navigation/NavigationService');
/**
 * Tests
 */
describe('Messenger setup component', () => {
  let activity;

  beforeEach(() => {
    activity = activitiesServiceFaker().load(1).activities[0];
  });

  it('should renders correctly with source', async () => {
    const video = renderer
      .create(
        <MindsVideo source={{ uri: 'https//www.minds.com/somevidep.mp4' }} />,
      )
      .toJSON();

    expect(video).toMatchSnapshot();
  });

  it('should renders correctly with entity', async () => {
    const video = renderer
      .create(<MindsVideo entity={ActivityModel.create(activity)} />)
      .toJSON();

    expect(video).toMatchSnapshot();
  });
});
