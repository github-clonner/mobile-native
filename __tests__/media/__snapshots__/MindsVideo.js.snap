// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messenger setup component should renders correctly with entity 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "black",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "flexDirection": "column",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "alignContent": "center",
            "alignItems": "center",
            "alignSelf": "center",
            "justifyContent": "center",
          },
          Object {
            "marginTop": 10,
          },
        ]
      }
    >
      <Text
        allowFontScaling={false}
        onPress={[Function]}
        style={
          Array [
            Object {
              "color": "#f7f7f7",
              "fontSize": 56,
            },
            Object {
              "alignSelf": "center",
              "bottom": 0,
              "left": 0,
              "position": "relative",
              "right": 0,
              "top": 0,
            },
            Object {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`Messenger setup component should renders correctly with source 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "black",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "flexDirection": "column",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "alignContent": "center",
            "alignItems": "center",
            "alignSelf": "center",
            "justifyContent": "center",
          },
          Object {
            "marginTop": 10,
          },
        ]
      }
    >
      <Text
        allowFontScaling={false}
        onPress={[Function]}
        style={
          Array [
            Object {
              "color": "#f7f7f7",
              "fontSize": 56,
            },
            Object {
              "alignSelf": "center",
              "bottom": 0,
              "left": 0,
              "position": "relative",
              "right": 0,
              "top": 0,
            },
            Object {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "stretch",
          "backgroundColor": "rgba(48,48,48,0.7)",
          "borderRadius": 3,
          "bottom": 0,
          "flexDirection": "row",
          "left": 0,
          "margin": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "padding": 5,
            },
            Object {
              "flexDirection": "row",
              "justifyContent": "space-evenly",
            },
            Object {
              "marginRight": 5,
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          onPress={[Function]}
          style={
            Array [
              Object {
                "color": "#f7f7f7",
                "fontSize": 23,
              },
              Object {
                "paddingLeft": 5,
              },
              Object {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "height": 3,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "white",
                  },
                  Object {
                    "marginRight": 10,
                  },
                ]
              }
            >
              00:00
            </Text>
            <View
              onLayout={[Function]}
              onMoveShouldSetResponder={[Function]}
              onMoveShouldSetResponderCapture={[Function]}
              onResponderEnd={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderReject={[Function]}
              onResponderRelease={[Function]}
              onResponderStart={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              onStartShouldSetResponderCapture={[Function]}
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "height": 3,
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                isTVSelectable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "borderColor": "#4690D6",
                    "borderWidth": 3,
                    "flex": 0,
                    "opacity": 1,
                    "padding": 0,
                  }
                }
              />
              <View
                accessible={true}
                focusable={true}
                isTVSelectable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "borderColor": "#f7f7f7",
                    "borderWidth": 3,
                    "flex": 100,
                    "opacity": 1,
                    "padding": 0,
                  }
                }
              />
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "white",
                  },
                  Object {
                    "marginLeft": 10,
                  },
                ]
              }
            >
              00:00
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "padding": 5,
            },
            Object {
              "flexDirection": "row",
              "justifyContent": "space-evenly",
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          onPress={[Function]}
          style={
            Array [
              Object {
                "color": "#f7f7f7",
                "fontSize": 23,
              },
              undefined,
              Object {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;
